zerver/lib/push_notifications.py

```diff
@@ -547,6 +547,42 @@ def get_gcm_payload(user_profile: UserProfile, message: Message) -> Dict[str, An
     })
     return data
 
+def handle_remove_push_notification(user_profile_id: int, remove_message: Dict[str, Any]) -> None:
+    """
+    Sends a GCM message to the mobile app when the message is read on the server, to remove the message
+    from the notification.
+    """
+
+    user_profile = get_user_profile_by_id(user_profile_id)
+    message = access_message(user_profile, remove_message['message_id'])[0]
+    gcm_payload = get_common_payload(message)
+    content = truncate_content(get_mobile_push_content(message.rendered_content))[0]
+
+    gcm_payload.update({
+        'user': user_profile.email,
+        'event': 'remove',
+        'sender_full_name': message.sender.full_name,
+        'content': content,
```

Comment:
Why are we sending the content, not a message ID?

Reply:
The ID's aren't being stored in the DS which is used to save the notification content history, so to find if it's the last message in the notifcation history then the content is being used.

This is the DS used in Java to save the content of the last message receieved in each narrow.
 ```
LinkedHashMap<String, Pair<String, Integer>> conversations;
```
Here the string is keyword of the narrow, and pair contains the content of last message and the number of messages in that narrow.

Reply:
Is it possible to change the model in the app to store the message IDs?  It's clear that message IDs are the right data key for this (well, server identity + message ID, since we need to think about the case that there might be multiple servers involved).

Actually ... with the push bouncer, does this protocol currently have a way to communicate which server the notice is related to?  We'll want that to support multiple servers being active at the same time...

Reply:
("Content" is messy as a protocol, because it's subject to editing, so to be correct we'd need to compute what the content was at the time the notification went out to the user, which is annoying)

Reply:
Yes, The notification style currently is configured for just showing the data, which worked well till now. And there was no requirement to store the message ID.

Now, I'll create a new class`LastMessageInfo` and have data like `MessageId`, `MessageContent`, `TotalMessageCount`.

So we can have a DS like 
``` 
LinkedHashMap<String, LastMessageInfo> conversations;
```

Reply:
Yes, there is a realm string being sent from here so we can detect different servers.

Reply:
Yeah, we definitely want this to be in terms of message IDs; specifically, message IDs namespaced by server.

(If that means this feature won't work on existing versions of the mobile app, that's fine; people update pretty fast.)

We'll want to keep more of the information than that on the client side in any case; e.g. when I get 3 messages in a row in a given PM conversation, I typically want the notification area to show me all 3, not only the latest.

It would be good to straighten out how our push notification protocol in general handles namespacing things (including user IDs too) by server. I haven't looked closely at this area recently, but I remember having the sense when I first looked at it late last summer that it wasn't clearly consistent about that.

I think one acceptable way to identify a "server" here would be the URL of an org/realm. In principle there are optimizations we could do by the client knowing that several orgs are all the same server underneath; but those aren't for the near or medium term, I think (and there's careful server-side work we'd have to do for them, and I don't think they'd do much for notifications anyway.) The main thing then is to write down a really clear convention for what form of the URL to use, e.g. whether it includes the `https://`.

Reply:
So I think one reasonable payload to send here would look like
```
{
  'realm': ... # ? This needs work in the new-message case, too
  'user': user_profile.email,
  'event': 'remove_message',
  'zulip_message_id': message.id,
  # ... then something to say that the system shouldn't display this notification itself?
  # I haven't looked up the API for this.
}
```

Reply:
I think `realm` should be the canonical hostname for the realm, i.e. `realm.uri`.  That should be both (1) uniquely mapped to a realm and (2) stable enough -- I think it's expected that if a server's hostname changes, you'd need to re-login to mobile apps for push notifications (you'd need to do so anyway to not have the API reject all your other requests).

Reply:
> We'll want to keep more of the information than that on the client side in any case; e.g. when I get 3 messages in a row in a given PM conversation, I typically want the notification area to show me all 3, not only the latest.

So i'll replicate like whatsapp does, which is to have two level of expandable notification, the first one is similar to our app, showing the number of messages and show the last message message.

And when this thing is expanded each conversation expands to its own notification which can be further expanded for more detailed view. So here I would store the message ID's as well.

Reply:
Now multiple notifications are being sent, previously it was one and was being edited.
And in place of the email (janedoe@example) we can show name or email of receiver account for multiple accounts, and we can show the realm name as well in place of Notification test? 
 
Screenshots: 
Compressed
[device-2018-08-01-191329](https://user-images.githubusercontent.com/12700799/43525339-26379668-95bf-11e8-9280-d7297d26ca4f.png)

One level of expansion 
[device-2018-08-01-191342](https://user-images.githubusercontent.com/12700799/43525340-268978ac-95bf-11e8-8384-3922b6c0adc3.png)

On expanding the notification
[device-2018-08-01-191357](https://user-images.githubusercontent.com/12700799/43525341-26bb11b4-95bf-11e8-83d9-55de213b2c60.png)

Reply:
Yeah, a lot like those screenshots. I'd show the sender's name instead of their email address, but that's orthogonal to this change.

Here's an example of the FB Messenger behavior I had in mind (from a [random article](http://www.ubergizmo.com/2016/06/facebook-messenger-android-quick-reply/) I found by googling a bit):
![image](https://user-images.githubusercontent.com/28173/43528244-65149c9e-955d-11e8-84fa-f33eade8c791.png)

On the right-hand side, the way it shows the whole list of messages -- I often find that super helpful when I get a series of messages from someone.

Reply:
@timabbott 
> I think `realm` should be the canonical hostname for the realm, i.e. realm.uri (aka not including the https://.

Sounds good to me!

What about ports? I suppose if someone's running their server on an alternate port, it'd be best not to just unnecessarily break by forgetting that; so would this canonical name then be something like `chat.example.edu:444`?

It'd be good for this to correspond to something explicit in the API. Browsing a bit, a full `/register` response has these fields:
```
  "realm_bot_domain": "kandra-test.zulipchat.com",
  "realm_uri": "https://kandra-test.zulipchat.com",
```

I haven't traced through source to confirm exactly what those mean. Do you expect that `realm_bot_domain` should be identical to the canonical `realm` value you have in mind?

Browsing a bit more, there's no value like that in a `/server_settings` response; though there is `realm_uri` again, including the `https://`.

Reply:
Submitted https://github.com/zulip/zulip-mobile/pull/2855 to save the message id in Android

Reply:
Oops, I guess I'd meant `realm.host` above, but `realm_uri` seems fine (I hadn't had a chance to look at how it's exposed in the API currently).

Reply:
It appears we already have these two elements in `get_common_payload`, so it is already possible to distinguish the realm in the notifications:

```
    data['server'] = settings.EXTERNAL_HOST                                                                     
    data['realm_id'] = message.sender.realm.id                                                                  
```

(And we should use that in fallback code in the app if `realm_uri` isn't there).  But I think it makes sense to add `realm_uri`, since that lines up properly with the rest.

Reply:
For message IDs, they are already present in the APNS notices in some fashion due to this:
```
def get_apns_payload(message: Message) -> Dict[str, Any]:                                                       
    zulip_data = get_common_payload(message)                                                                    
    zulip_data.update({                                                                                         
        'message_ids': [message.id],                                                                            
    })                                                                                                          
```

I'd need to understand how APNS works better to know whether that's good, but certainly for GCM, I think we want to add

Reply:
I merged 5f8d193bb7cf137647db265092fb079aca5719fc for the `realm_uri` piece.

Reply:
What would be the best way to show the narrow information and how?
Would it be something like this? 

[device-2018-08-05-154045](https://user-images.githubusercontent.com/12700799/43684871-f3257362-98c5-11e8-9a5c-168bc0077f5f.png)

Basically in the left of time indicator?

Reply:
Hmm, I'm not sure, that's probably a discussion for the display-logic side PR